Waypoint generation: grid rotation, aoi clipping, no-fly zone handling

azharcodeit opened the pull request:
This PR enhances the flight plan waypoint generator by introducing features for arbitrary waypoint grid rotation and precise clipping to the Area of Interest (AOI).
Resolves issue #7 

## Key Changes:

### **Arbitrary Rotation of Waypoints:**

-   **Previous Behavior:** Waypoints were generated in a grid aligned strictly along the east-west axis.
-   **New Implementation:** A `rotation_angle` parameter has been introduced, allowing users to specify the angle at which the waypoint grid should be generated. The grid is now rotated around the centroid of the AOI according to this angle, ensuring waypoints align with the specified orientation.
-   **Technical Details:** The `shapely.affinity.rotate` function is used to rotate both the AOI and the generated grid points. After the waypoints are generated, they are rotated back to their original coordinate system to ensure accurate placement.


### **AOI Clipping:**

-   **Previous Behavior:** Waypoints were generated within the bounding box of the AOI, without considering the exact AOI boundaries.
-   **New Implementation:** The grid of waypoints is now clipped to the precise shape of the AOI, ensuring that all generated waypoints lie within the intended area, avoiding unnecessary or invalid waypoints outside the AOI.
-   **Technical Details:** Waypoints are included only if they fall within the rotated AOI, ensuring precise coverage of the area of interest.


### **Handling No-Fly Zones:**

-   **Previous Behavior:** The waypoint generator did not account for no-fly zones within the AOI.
-   **New Implementation:** The generator now excludes waypoints that fall within no-fly zones, ensuring that waypoints are not generated in restricted areas.
-   **Limitations:** The current implementation does not reroute or adjust the path to navigate around no-fly zones. Instead, it simply excludes waypoints that would fall inside these zones, which may result in gaps in the waypoint grid.

## Output: 
  **GeoJSON Output:** The final output is a` GeoJSON FeatureCollection` where each `Feature` represents a waypoint with the following:
  Geometry: A `Point` with coordinates (x, y)

```
Example output:
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          -0.130663,
          51.525383
        ]
      },
      "properties": {
        "index": 0,
        "heading": -90,
        "take_photo": true,
        "gimbal_angle": "-90"
      }
    },
```

## Testing:

-   **Visual Testing:** The rotation and clipping functionalities were visually tested using various regular and irregular shapes for rotation angles of 0°, 45°, 90° and 120°. 

<img width="180" alt="Screenshot 2024-09-02 at 18 25 28" src="https://github.com/user-attachments/assets/1d087913-a1f9-4a7e-9ba4-45f5e622f120">
<img width="180" alt="Screenshot 2024-09-02 at 18 26 13" src="https://github.com/user-attachments/assets/d47aa6c6-42bf-4931-944d-6963be4ec5bd">
<img width="180" alt="Screenshot 2024-09-02 at 18 26 27" src="https://github.com/user-attachments/assets/0c4f89da-55f0-4d48-b19d-b510bd8a546b">
<img width="180" alt="Screenshot 2024-09-02 at 18 26 51" src="https://github.com/user-attachments/assets/a9172952-fa80-4385-8b76-2ccebca11859">

<img width="180" alt="Screenshot 2024-09-02 at 18 20 41" src="https://github.com/user-attachments/assets/a47ba368-ea15-4610-8961-dc3729af5428">
<img width="180" alt="Screenshot 2024-09-02 at 18 21 12" src="https://github.com/user-attachments/assets/d6cce386-5cc8-4600-ac5f-92541916e44f">
<img width="180" alt="Screenshot 2024-09-02 at 18 22 12" src="https://github.com/user-attachments/assets/33127834-1b28-451a-9dfa-4ffb35199204">
<img width="180" alt="Screenshot 2024-09-02 at 18 39 04" src="https://github.com/user-attachments/assets/158f355c-8d01-4221-96e1-647fd969e2b3">






-   **Rotation Accuracy:** Confirmed that the waypoints correctly align with the specified rotation angles, maintaining the desired grid structure.
-   **AOI Clipping:** Verified that all waypoints are contained within the AOI boundary and correctly exclude points outside the defined area.
-   **No-Fly Zone Handling:** Ensured that waypoints within specified no-fly zones were appropriately excluded from the final output.

**Notes:**
  The rotation logic was carefully reviewed and tested through visualizations to ensure correct alignment of waypoints at various angles.

**Example Command:**

```
python3 waypoints.py \
--project_geojson_polygon path/to/aoi.geojson \
--altitude_above_ground_level 120 \
--forward_overlap 80.0 \
--side_overlap 80.0 \
--rotation_angle 90 \
--output_file_path path/to/output.geojson \
--no_fly_zones path/to/nfz.geojson
```

nrjadkry commented:
Thanks @azharcodeit , This is nice. Well done.

Just a few points:
- I had created the way-points in the bounding box of the polygon, you have remove them and have the waypoints just inside the polygon. This is nice and should be like this. 
- I had added an extra point outside the bounding box and had the drone not take a picture at that point. It was done because drone cannot make a sharp 90 degree turn and makes a curve like shape turn because of which, the picture taken at that point won't be of any use.  So, I suggest you to add an extra point just outside the polygon.
- No fly zone means that the drone should not fly above above that area. I know that it is tricky to  handle this properly. But, just not having any waypoints in the no fly zone does not make sense.

nrjadkry commented:
Thanks, @azharcodeit. This is excellent work—well done!

A few suggestions:

- I initially created waypoints within the bounding box of the polygon, but you've modified this to have the waypoints only within the polygon itself. This change is great and should remain as is.

- I had previously added an extra point just outside the bounding box where the drone was not supposed to take a picture. This was done because the drone cannot execute a sharp 90-degree turn and instead makes a curved turn. As a result, the picture taken at that point wouldn't have been useful. I recommend adding a similar extra point just outside the polygon to account for this.

- Regarding the no-fly zone: the drone should not fly over these areas at all. I understand that handling this correctly can be challenging, but simply avoiding waypoints within the no-fly zone isn't enough. There needs to be a more comprehensive solution.

spwoodcock commented:
Thanks for the great feedback Niraj!

After discussion we decided:

- The second point should be addressed by modifying this PR.

- The third point should be addressed in another follow up PR. Based on timings we will decide if @azharcodeit has time to do this or if we hand over to someone else to take it up (as the internship period has technically ended now)

azharcodeit commented:
Thank you, @nrjadkry and @spwoodcock, for the feedback!

I've addressed the second point by adding extra start and end waypoints outside the polygon for each row, ensuring the drone doesn't take photos during turns. I've also updated the visualizations in the PR description to reflect these changes, which are included in the latest commit.

Regarding the no-fly zones, as mentioned in the PR description, I've updated the waypoint generator to exclude waypoints within these zones. However, I acknowledge that this is a starting point, and a more robust solution is needed to prevent the drone from flying over no-fly zones when traveling between waypoints. I'll be working on this in a follow-up PR, as discussed.

Thanks again for your guidance and suggestions!

nrjadkry commented:
@azharcodeit, One more thing, I apologize for not mentioning this earlier. I added two points in each line (instead of one) after the generated ones . The first point captures a picture, while the second does not. This was done to ensure that no areas are left unmapped, as some edge regions were found to be missing, generally in the right side of the polygon.

In this case, you've added the buffer distance, which sometimes falls inside the polygon ( when the buffer distance is less than the forward_overlap). Not capturing an image at that point could result in missing some areas of the polygon. To address this, can you add forward_spacing * 2, capturing an image at the first point while skipping the second poin ?

![image](https://github.com/user-attachments/assets/07b2a4a7-0971-4d5f-8823-a469b7395ff2)

azharcodeit commented:
Thanks for the feedback, @nrjadkry! No worries, everything makes sense.

To achieve maximum coverage, I initially added 2 points instead of one, but in certain cases, this approach didn’t work well and left some uncovered space. To address this, I propose adding a whole new line of waypoints that are close to or on the edges. After several trials, I found this to be more effective and it didn't require much expansion of the code.

For example:
- **First Image**: This shows the result of adding 2 points.
- **Second Image**: This shows adding a new waypoints near to the edges. (Additional cases are attached in PR description)

Please let me know if proposed change align with your expectations.

<img width="250" alt="Screenshot 2024-09-02 at 19 07 46" src="https://github.com/user-attachments/assets/c9dd1e19-5620-4b4f-83f7-7e67f0127ee6">

<img width="250" alt="Screenshot 2024-09-02 at 18 25 28" src="https://github.com/user-attachments/assets/9477682a-cf38-494c-b6d2-254fcb730b6f">

spwoodcock commented on drone_flightplan/create_flightplan.py:
```diff
@@ -84,7 +91,7 @@ def main():
         "--gsd",
         type=float,
-        help="The flight altitude in meters.",
+        help="The ground sampling distance in meters.",
```
Thanks for the correction here! Units are actually `cm/px` (centimetres per pixel)

azharcodeit replied:
Thanks, @spwoodcock! I've corrected the unit measurements

spwoodcock commented on drone_flightplan/waypoints.py:
```diff
@@ -2,345 +2,147 @@
-    y = miny
-    row_count = 0
-    angle = -90
+def generate_grid_in_aoi(aoi_polygon, x_spacing, y_spacing):
```
Although not essential, type hinting is a really useful feature in Python - one of the understated advantages is to make it easier to understand code when looking back in the future!

```python
def generate_grid_in_aoi(aoi_polygon: shape, x_spacing: int, y_spacing: int):
```

This tells the current reviewer, future you, future collaborator that the aoi_polygon should be a shapely shape 👍

Of course, type hinting can also be used to validate params and do other things too!

azharcodeit replied:
Appreciate the suggestion! I've added type hinting to all relevant functions in the latest commit.

spwoodcock commented on drone_flightplan/waypoints.py:
```diff
@@ -2,345 +2,147 @@
+    return points
+
+
+def create_path(points, forward_spacing, generate_3d=False):
```
Docstrings are also far from essential, but a simple description of what a function does can really help in the future!

```python
def create_path(...):
    """Given a grid of points, creates a wayline path between them for a drone to fly."""
    ...
```

azharcodeit replied:
Thanks! I've now included docstrings in the functions.

spwoodcock reviewed (commented):
Nice work!

spwoodcock reviewed (approved):
Nice update! Looks excellent!

I'll let Niraj have the final say 😄

nrjadkry reviewed (approved):
Nice Work. This is ready to be merged. 🎉

nrjadkry commented:
@azharcodeit Looks like `generate_each_points` in the waypoints is not doing anything. We create only the boundary point when it is False. Have you created any other function to do that ?

azharcodeit commented:
Hi @nrjadkry,

Thanks for pointing that out! 

To clarify, are we aiming for it to generate `waylines` instead of individual `waypoints` when `generate_each_points=False`? If so, I'll adjust the logic to ensure that it generates waylines.

Let me know your thoughts, and I can update the implementation accordingly.

nrjadkry commented:
Yes, exactly. When we have generate_each_points=False, we would want waylines.
Also, I can see that there are multiple points generated at the same point when we have generate_each_points set to True. Will you also look into that ?

azharcodeit commented:
@nrjadkry, thanks for confirming.

I’ll investigate both cases, so that:

`generate_each_points = False` => generates `waylines`
`generate_each_points = True` => generates `waypoints`
I’ll address these issues and follow up with a PR.

Thanks!
